x-pack/platform/plugins/shared/fleet/server/services/agentless/deployment_sync.ts

```diff
@@ -0,0 +1,191 @@
+import { AgentlessAgentListNotFoundError } from '../../errors';
+
+const AGENTLESS_CONCURRENCY = 1;
+const PAGE_SIZE = 20;
+
+function dryRunTag(dryRun = false) {
+  if (dryRun) {
+    return `[Dry Run]`;
+  }
+  return '';
+}
+
+export async function syncAgentlessDeployments(
+  {
+    logger,
+    agentlessAgentService,
+  }: {
+    logger: Logger;
+    agentlessAgentService: AgentlessAgentService;
+  },
+  opts?: {
+    dryRun?: boolean;
+    abortController?: AbortController;
+  }
+) {
+  logger.info(`[Agentless Deployment Sync] Starting sync process`);
+  const soClient = appContextService.getInternalUserSOClientWithoutSpaceExtension();
+
+  try {
+    // retrieve all agentless deployments
+    const currentDeployments: Array<{ policy_id: string; revision_idx?: number }> = [];
+    let hasMore = true;
+    let nextPageToken: string | undefined;
+
+    while (hasMore) {
+      if (opts?.abortController?.signal.aborted) {
+        throw new Error('Task was aborted');
+      }
+      const deploymentRes = await agentlessAgentService.listAgentlessDeployments({
+        perPage: PAGE_SIZE,
+        nextPageToken,
+      });
+      if (!deploymentRes) {
+        return;
+      }
+
+      nextPageToken = deploymentRes.next_token;
+      if (!nextPageToken) {
+        hasMore = false;
+      }
+
+      const policyIds = deploymentRes.deployments.map((d) => d.policy_id);
+      const agentPolicies = await agentPolicyService.getByIds(soClient, policyIds, {
+        spaceId: '*',
+        ignoreMissing: true,
+      });
+
+      await pMap(
+        deploymentRes.deployments,
+        async (deployment) => {
+          if (opts?.abortController?.signal.aborted) {
+            throw new Error('Task was aborted');
+          }
+          const agentPolicy = agentPolicies.find((ap) => ap.id === deployment.policy_id);
+
+          if (!agentPolicy) {
+            logger.info(
+              `[Agentless Deployment Sync]${dryRunTag(opts?.dryRun)} Deleting deployment ${
+                deployment.policy_id
+              }`
+            );
+            if (!opts?.dryRun) {
+              await agentlessAgentService
+                .deleteAgentlessAgent(deployment.policy_id)
+                .catch((error) => {
+                  logger.error(
+                    `[Agentless Deployment Sync] Failed to delete deployment ${deployment.policy_id}`,
+                    { error }
+                  );
+                });
+            }
+          } else if (
+            deployment.revision_idx === undefined ||
+            deployment.revision_idx < agentPolicy.revision
+          ) {
+            logger.info(
+              `[Agentless Deployment Sync]${dryRunTag(opts?.dryRun)} Updating deployment ${
+                deployment.policy_id
+              }`
+            );
+            if (!opts?.dryRun) {
+              const spacedScoppedSoClient = appContextService.getInternalUserSOClientForSpaceId(
```

Comment:
Question: Do the clients in this function need to use the `InternalUser` variants?

Reply:
This code is mean to run as an automated task it's why we are using an internal client for saved objects. For the ES client we use an internal one too to access `.fleet` system indices, does it make things clearer?